### PR TITLE
feat(es/minifier): Respect more options

### PR DESCRIPTION
**Pull request by kdy1**
**Description:**

This PR makes the ES minifier respect more options.

Changes:
 - `reduce_vars` now enables IIFE invoker.
 - `reduce_vars` now enables seq inliner.
 - `top_retain` is now respected by seq inliner.
 - `arguments` is now ignored by seq inliner.

### CODE REVIEW

**Review by kdy1 — COMMENTED** [review]
---
swc-bump:
 - swc_ecma_minifier

File: crates/swc_ecma_minifier/src/compress/optimize/sequences.rs [hunk 1]
```diff
@@ -1513,6 +1522,15 @@ impl Optimizer<'_> {
             )
         };
 
+        // Respect top_retain
+        if let Some(a_id) = a.id() {
+            if a_id.0 == "arguments"
+                || (self.ctx.in_top_level() && self.options.top_retain.contains(&a_id.0))
```
Comment (Austaras) [inline crates/swc_ecma_minifier/src/compress/optimize/sequences.rs:1528]: Maybe just use `may_remove_ident`

Comment (kdy1) [reply crates/swc_ecma_minifier/src/compress/optimize/sequences.rs]: `if a_id.0 == "arguments" || !self.may_remove_ident(&Ident::from(a_id)) {` causes lots of diffs.

```
diff --git a/crates/swc_ecma_minifier/tests/fixture/projects/backbone/12/output.js b/crates/swc_ecma_minifier/tests/fixture/projects/backbone/12/output.js
index bf7e3b0e96..680bbfce0f 100644
--- a/crates/swc_ecma_minifier/tests/fixture/projects/backbone/12/output.js
+++ b/crates/swc_ecma_minifier/tests/fixture/projects/backbone/12/output.js
@@ -1,4 +1,4 @@
 var names, i, l;
-for(i = 0, l = (names = name ? [
+for(i = 0, names = name ? [
     name
-] : _.keys(this._events)).length; i < l; i++)name = names[i];
+] : _.keys(this._events), l = names.length; i < l; i++)name = names[i];
diff --git a/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/cascade_forin/output.js b/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/cascade_forin/output.js
index d19450d3cc..4f45d6982f 100644
--- a/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/cascade_forin/output.js
+++ b/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/cascade_forin/output.js
@@ -1,5 +1,9 @@
 var a;
 function f(b) {
-    return [b, b, b];
+    return [
+        b,
+        b,
+        b
+    ];
 }
-for (var c in f((a = console))) console.log(c);
+for(var c in a = console, f(a))console.log(c);
diff --git a/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_1631_1/output.js b/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_1631_1/output.js
index 75c146598b..dc7654d1c1 100644
--- a/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_1631_1/output.js
+++ b/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_1631_1/output.js
@@ -1,9 +1,9 @@
 function f(x) {
-    return (pc = 200), 100;
+    return pc = 200, 100;
 }
 function x() {
     var t = f();
-    return (pc += t);
+    return pc += t, pc;
 }
 var pc = 0;
 console.log(x());
diff --git a/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2954_2/output.js b/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2954_2/output.js
index 3388d171ea..05367f4ee6 100644
--- a/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2954_2/output.js
+++ b/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2954_2/output.js
@@ -3,10 +3,11 @@ try {
     throw 0;
 } catch (e) {
     do {
-        a = "FAIL_2";
-        (b = function() {
+        b = function() {
             throw Error("PASS");
-        }()) && b.c;
+        }();
+        a = "FAIL_2";
+        b && b.c;
     }while (0)
 }
 console.log(a);
diff --git a/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2954_3/output.js b/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2954_3/output.js
index 8d7012c8c0..cf9f48d07d 100644
--- a/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2954_3/output.js
+++ b/crates/swc_ecma_minifier/tests/terser/compress/collapse_vars/issue_2954_3/output.js
@@ -1,10 +1,11 @@
 var b, a = "FAIL_1";
 try {} finally{
     do {
-        a = "FAIL_2";
-        (b = function() {
+        b = function() {
             throw Error("PASS");
-        }()) && b.c;
+        }();
+        a = "FAIL_2";
+        b && b.c;
     }while (0)
 }
 console.log(a);
diff --git a/crates/swc_ecma_minifier/tests/terser/compress/functions/issue_2601_1/output.js b/crates/swc_ecma_minifier/tests/terser/compress/functions/issue_2601_1/output.js
index 9e44326bad..f0c9ffe6b1 100644
--- a/crates/swc_ecma_minifier/tests/terser/compress/functions/issue_2601_1/output.js
+++ b/crates/swc_ecma_minifier/tests/terser/compress/functions/issue_2601_1/output.js
@@ -1,2 +1,2 @@
 var a = "FAIL";
-console.log(a = "PASS");
+a = "PASS", console.log(a);
diff --git a/crates/swc_ecma_minifier/tests/terser/compress/functions/issue_2620_1/output.js b/crates/swc_ecma_minifier/tests/terser/compress/functions/issue_2620_1/output.js
index 81e5626c69..ad41245f97 100644
--- a/crates/swc_ecma_minifier/tests/terser/compress/functions/issue_2620_1/output.js
+++ b/crates/swc_ecma_minifier/tests/terser/compress/functions/issue_2620_1/output.js
@@ -1,2 +1,2 @@
 var c = "FAIL";
-console.log(c = "PASS");
+c = "PASS", console.log(c);
diff --git a/crates/swc_ecma_minifier/tests/terser/compress/loops/dead_code_condition/output.js b/crates/swc_ecma_minifier/tests/terser/compress/loops/dead_code_condition/output.js
index a33152c6e7..a1cbd3fe80 100644
--- a/crates/swc_ecma_minifier/tests/terser/compress/loops/dead_code_condition/output.js
+++ b/crates/swc_ecma_minifier/tests/terser/compress/loops/dead_code_condition/output.js
@@ -1,3 +1,3 @@
 var a = 0, b = 5;
 var c;
-console.log(a += 1);
+a += 1, console.log(a);
diff --git a/crates/swc_ecma_minifier/tests/terser/compress/sequences/func_def_1/output.js b/crates/swc_ecma_minifier/tests/terser/compress/sequences/func_def_1/output.js
index 741651e9cc..4084b304f5 100644
--- a/crates/swc_ecma_minifier/tests/terser/compress/sequences/func_def_1/output.js
+++ b/crates/swc_ecma_minifier/tests/terser/compress/sequences/func_def_1/output.js
@@ -1,4 +1,4 @@
 function f() {
-    return f = 0, false;
+    return f = 0, !!f;
 }
 console.log(f());
diff --git a/crates/swc_ecma_minifier/tests/terser/compress/sequences/lift_sequences_2/output.js b/crates/swc_ecma_minifier/tests/terser/compress/sequences/lift_sequences_2/output.js
index 0cd3bb6045..77e27bc052 100644
--- a/crates/swc_ecma_minifier/tests/terser/compress/sequences/lift_sequences_2/output.js
+++ b/crates/swc_ecma_minifier/tests/terser/compress/sequences/lift_sequences_2/output.js
@@ -1,2 +1,2 @@
 var foo = 1, bar;
-foo.x = (foo = {}, 10), bar = {}, console.log(foo, bar = 10);
+foo.x = (foo = {}, 10), bar = {}, bar = 10, console.log(foo, bar);
diff --git a/crates/swc_ecma_minifier/tests/terser/compress/sequences/lift_sequences_5/output.js b/crates/swc_ecma_minifier/tests/terser/compress/sequences/lift_sequences_5/output.js
index 6a3efccc7b..e450a86637 100644
--- a/crates/swc_ecma_minifier/tests/terser/compress/sequences/lift_sequences_5/output.js
+++ b/crates/swc_ecma_minifier/tests/terser/compress/sequences/lift_sequences_5/output.js
@@ -1,2 +1,2 @@
 var a = 2, b;
-console.log(a *= (a = 4, 3));
+a *= (a = 4, 3), console.log(a);

```

Comment (kdy1) [reply crates/swc_ecma_minifier/src/compress/optimize/sequences.rs]: I think `may_remove_ident` should return `true` at the end,or 
```rust
        if self.options.top_level() {
            !self.options.top_retain.contains(&id.sym)
        } else {
            !self.ctx.in_top_level()
        }
```

Comment (Austaras) [reply crates/swc_ecma_minifier/src/compress/optimize/sequences.rs]: Hmm. Maybe `if a_id.0 == "arguments" || (matches!(a, Mergable::Var(_)) && !self.may_remove_ident(&Ident::from(a_id)))`

Comment (kdy1) [reply crates/swc_ecma_minifier/src/compress/optimize/sequences.rs]: It still regresses. 
image

```diff
diff --git a/crates/swc/tests/vercel/full/react-hooks/1/output/index.js b/crates/swc/tests/vercel/full/react-hooks/1/output/index.js
index 1e7b1517d2..9075213fe6 100644
--- a/crates/swc/tests/vercel/full/react-hooks/1/output/index.js
+++ b/crates/swc/tests/vercel/full/react-hooks/1/output/index.js
@@ -4,8 +4,8 @@ import { useRouter as e } from "next/router";
 import { useProject as o } from "@swr/use-project";
 import a from "@swr/use-team";
 export default function m() {
-    var m = o(e().query.project).data;
+    var m = e().query, u = m.project, s = o(u), i = s.data;
     a().teamSlug;
-    var u = null == m ? void 0 : m.id;
-    return useProjectBranches(u).data, r(t, {});
+    var n = null == i ? void 0 : i.id, p = useProjectBranches(n);
+    return p.data, r(t, {});
 }
diff --git a/crates/swc_ecma_minifier/src/compress/optimize/sequences.rs b/crates/swc_ecma_minifier/src/compress/optimize/sequences.rs
index ea189114b7..ce1948b28d 100644
--- a/crates/swc_ecma_minifier/src/compress/optimize/sequences.rs
+++ b/crates/swc_ecma_minifier/src/compress/optimize/sequences.rs
@@ -1525,9 +1525,7 @@ impl Optimizer<'_> {
         // Respect top_retain
         if let Some(a_id) = a.id() {
             if a_id.0 == "arguments"
-                || (self.ctx.in_top_level()
-                    && a_id.1 == self.marks.top_level_ctxt
-                    && self.options.top_retain.contains(&a_id.0))
+                || (matches!(a, Mergable::Var(_)) && !self.may_remove_ident(&Ident::from(a_id)))
             {
                 return Ok(false);
             }
diff --git a/crates/swc_ecma_minifier/tests/fixture/issues/5280/output.js b/crates/swc_ecma_minifier/tests/fixture/issues/5280/output.js
index b0173968af..c154de5a85 100644
--- a/crates/swc_ecma_minifier/tests/fixture/issues/5280/output.js
+++ b/crates/swc_ecma_minifier/tests/fixture/issues/5280/output.js
@@ -1,3 +1,4 @@
 export function source() {
-    console.log(6, 1, 1);
+    let c = 0, a = 1, b = c += a;
+    console.log(a += 5, b, c);
 }
diff --git a/crates/swc_ecma_minifier/tests/fixture/issues/emotion/react/2/output.js b/crates/swc_ecma_minifier/tests/fixture/issues/emotion/react/2/output.js
index afe5929aa6..2bdb7aa106 100644
--- a/crates/swc_ecma_minifier/tests/fixture/issues/emotion/react/2/output.js
+++ b/crates/swc_ecma_minifier/tests/fixture/issues/emotion/react/2/output.js
@@ -108,19 +108,20 @@ export function serializeStyles(args, registered, mergedProps) {
     for(var i = 1; i < args.length; i++)styles += handleInterpolation(mergedProps, registered, args[i]), stringMode && (styles += strings[i]);
     labelPattern.lastIndex = 0;
     for(var identifierName = ""; null !== (match = labelPattern.exec(styles));)identifierName += "-" + match[1];
+    var name = function(str) {
+        for(var k, h = 0, i = 0, len = str.length; len >= 4; ++i, len -= 4)k = (0xffff & (k = 0xff & str.charCodeAt(i) | (0xff & str.charCodeAt(++i)) << 8 | (0xff & str.charCodeAt(++i)) << 16 | (0xff & str.charCodeAt(++i)) << 24)) * 0x5bd1e995 + ((k >>> 16) * 0xe995 << 16), k ^= k >>> 24, h = (0xffff & k) * 0x5bd1e995 + ((k >>> 16) * 0xe995 << 16) ^ (0xffff & h) * 0x5bd1e995 + ((h >>> 16) * 0xe995 << 16);
+        switch(len){
+            case 3:
+                h ^= (0xff & str.charCodeAt(i + 2)) << 16;
+            case 2:
+                h ^= (0xff & str.charCodeAt(i + 1)) << 8;
+            case 1:
+                h ^= 0xff & str.charCodeAt(i), h = (0xffff & h) * 0x5bd1e995 + ((h >>> 16) * 0xe995 << 16);
+        }
+        return h ^= h >>> 13, (((h = (0xffff & h) * 0x5bd1e995 + ((h >>> 16) * 0xe995 << 16)) ^ h >>> 15) >>> 0).toString(36);
+    }(styles) + identifierName;
     return {
-        name: function(str) {
-            for(var k, h = 0, i = 0, len = str.length; len >= 4; ++i, len -= 4)k = (0xffff & (k = 0xff & str.charCodeAt(i) | (0xff & str.charCodeAt(++i)) << 8 | (0xff & str.charCodeAt(++i)) << 16 | (0xff & str.charCodeAt(++i)) << 24)) * 0x5bd1e995 + ((k >>> 16) * 0xe995 << 16), k ^= k >>> 24, h = (0xffff & k) * 0x5bd1e995 + ((k >>> 16) * 0xe995 << 16) ^ (0xffff & h) * 0x5bd1e995 + ((h >>> 16) * 0xe995 << 16);
-            switch(len){
-                case 3:
-                    h ^= (0xff & str.charCodeAt(i + 2)) << 16;
-                case 2:
-                    h ^= (0xff & str.charCodeAt(i + 1)) << 8;
-                case 1:
-                    h ^= 0xff & str.charCodeAt(i), h = (0xffff & h) * 0x5bd1e995 + ((h >>> 16) * 0xe995 << 16);
-            }
-            return h ^= h >>> 13, (((h = (0xffff & h) * 0x5bd1e995 + ((h >>> 16) * 0xe995 << 16)) ^ h >>> 15) >>> 0).toString(36);
-        }(styles) + identifierName,
+        name: name,
         styles: styles,
         next: cursor
     };
diff --git a/crates/swc_ecma_minifier/tests/fixture/issues/react/hooks/1/output.js b/crates/swc_ecma_minifier/tests/fixture/issues/react/hooks/1/output.js
index 1622e0fc36..109e5532f2 100644
--- a/crates/swc_ecma_minifier/tests/fixture/issues/react/hooks/1/output.js
+++ b/crates/swc_ecma_minifier/tests/fixture/issues/react/hooks/1/output.js
@@ -3,8 +3,8 @@ import { useRouter } from "next/router";
 import { useProject } from "@swr/use-project";
 import useTeam from "@swr/use-team";
 export default function MyComp() {
-    var projectInfo = useProject(useRouter().query.project).data;
-    useTeam().teamSlug;
-    var projectId = null == projectInfo ? void 0 : projectInfo.id;
-    return useProjectBranches(projectId).data, _jsx(_Fragment, {});
+    var _query = useRouter().query, projectName = _query.project, ref = useProject(projectName), projectInfo = ref.data, ref1 = useTeam();
+    ref1.teamSlug;
+    var projectId = null == projectInfo ? void 0 : projectInfo.id, ref2 = useProjectBranches(projectId);
+    return ref2.data, _jsx(_Fragment, {});
 }
diff --git a/crates/swc_ecma_minifier/tests/fixture/issues/react/hooks/3/output.js b/crates/swc_ecma_minifier/tests/fixture/issues/react/hooks/3/output.js
index 0f318684bc..0747854032 100644
--- a/crates/swc_ecma_minifier/tests/fixture/issues/react/hooks/3/output.js
+++ b/crates/swc_ecma_minifier/tests/fixture/issues/react/hooks/3/output.js
@@ -6,13 +6,15 @@ __webpack_require__.d(__webpack_exports__, {
 });
 var router = __webpack_require__(86677), index_esm = __webpack_require__(45205), use_team = __webpack_require__(502), fetch_api = __webpack_require__(78869), authenticate = __webpack_require__(16966), api_endpoints = __webpack_require__(96236), qs = __webpack_require__(70326), use_project = __webpack_require__(76246);
 export function EnvironmentsSelector(param) {
-    var projectId, token, team, teamId, projectName = (0, router.useRouter)().query.project, projectInfo = (0, use_project.useProject)(projectName).data;
-    return (0, use_team.ZP)().teamSlug, (projectId = null == projectInfo ? void 0 : projectInfo.id, token = (0, authenticate.LP)(), teamId = null == (team = (0, use_team.ZP)().team) ? void 0 : team.id, (0, index_esm.ZP)(projectId ? "".concat(api_endpoints.Ms, "/git-branches").concat((0, qs.c)({
+    var token, team, teamId, _query = (0, router.useRouter)().query, projectName = _query.project, ref = (0, use_project.useProject)(projectName), projectInfo = ref.data, ref1 = (0, use_team.ZP)();
+    ref1.teamSlug;
+    var projectId = null == projectInfo ? void 0 : projectInfo.id, ref2 = (token = (0, authenticate.LP)(), teamId = null == (team = (0, use_team.ZP)().team) ? void 0 : team.id, (0, index_esm.ZP)(projectId ? "".concat(api_endpoints.Ms, "/git-branches").concat((0, qs.c)({
         projectId: projectId,
         teamId: teamId
     })) : "", function(endpoint) {
         return (0, fetch_api.Z)(endpoint, token, {
             throwOnHTTPError: !0
         });
-    }, void 0)).data, {};
+    }, void 0));
+    return ref2.data, {};
 }
diff --git a/crates/swc_ecma_minifier/tests/fixture/issues/react/hooks/4/output.js b/crates/swc_ecma_minifier/tests/fixture/issues/react/hooks/4/output.js
index d3209e8bc5..599bebc7a7 100644
--- a/crates/swc_ecma_minifier/tests/fixture/issues/react/hooks/4/output.js
+++ b/crates/swc_ecma_minifier/tests/fixture/issues/react/hooks/4/output.js
@@ -2,7 +2,7 @@
 __webpack_require__(86677);
 var index_esm = __webpack_require__(45205), use_team = __webpack_require__(502), fetch_api = __webpack_require__(78869), authenticate = __webpack_require__(16966), api_endpoints = __webpack_require__(96236), qs = __webpack_require__(70326);
 export function useProjectBranches(projectId, opts) {
-    var token = (0, authenticate.LP)(), team = (0, use_team.ZP)().team, teamId = null == team ? void 0 : team.id;
+    var token = (0, authenticate.LP)(), ref = (0, use_team.ZP)(), team = ref.team, teamId = null == team ? void 0 : team.id;
     return (0, index_esm.ZP)(projectId ? "".concat(api_endpoints.Ms, "/git-branches").concat((0, qs.c)({
         projectId: projectId,
         teamId: teamId
diff --git a/crates/swc_ecma_minifier/tests/full/issues/7094/1/output.js b/crates/swc_ecma_minifier/tests/full/issues/7094/1/output.js
index 5ce963deee..8bf1ee7bcf 100644
--- a/crates/swc_ecma_minifier/tests/full/issues/7094/1/output.js
+++ b/crates/swc_ecma_minifier/tests/full/issues/7094/1/output.js
@@ -1 +1 @@
-export function useSyncExternalStore$2(e,n,r){return n()}eval();
+export function useSyncExternalStore$2(e,t,n){let r=t();return r}eval();
diff --git a/crates/swc_ecma_minifier/tests/terser/compress/drop_unused/issue_t161_top_retain_3/output.js b/crates/swc_ecma_minifier/tests/terser/compress/drop_unused/issue_t161_top_retain_3/output.js
index 446e642654..2f276ce45e 100644
--- a/crates/swc_ecma_minifier/tests/terser/compress/drop_unused/issue_t161_top_retain_3/output.js
+++ b/crates/swc_ecma_minifier/tests/terser/compress/drop_unused/issue_t161_top_retain_3/output.js
@@ -1,4 +1 @@
-function f() {
-    return 2;
-}
-console.log(f(), 3);
+console.log(2, 3);

```

Comment (Austaras) [reply crates/swc_ecma_minifier/src/compress/optimize/sequences.rs]: Check my new commit. Should be fine now.

**Review by Austaras — APPROVED** [review]

**Pull request merged**